### PR TITLE
bugfix/20097-for-of-undefined-values

### PR DESCRIPTION
**Pull request by hubertkozik**
Changed for of loops to forEach to prevent undefined, touches #20097.

_____
I changed `for ... of` loops only for `series.data` array, because there is no possibility that `series.points` array will contain `undefined` values, so `for ... of` works perfectly there.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205924279035273

### CODE REVIEW

File: samples/unit-tests/series-errorbar/datalabels/demo.js [hunk 1]
```diff
@@ -1,29 +1,48 @@
...
+                [0, 19.6, 28.2],
+                [1, 18.1, 28.2],
+                [4, 10.8, 19.2]
             ]
-        }),
-        oldY = chart.series[0].points[1].dataLabelUpper.y;
+        }]
+    });
+
+    assert.ok(
```
[the remark this answers is not in the document]
Comment (bre1470) [reply samples/unit-tests/series-errorbar/datalabels/demo.js:29]: After testing a bit further it seems like errorbar behaves a bit strange. Taking xAxis.min = 3 and cropThreshold = 1 into account, I would expect that `series.data` contains two points (`undefined`, point of index 4) and `series.points` contains only the point. Maybe you can check with @KacperMadej whether this becomes a new issue for later.

**Review by bre1470 — APPROVED** [review]
Nice work! The test can be improved but otherwise it looks good. 👍

File: samples/unit-tests/series-errorbar/datalabels/demo.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hubertkozik) [reply samples/unit-tests/series-errorbar/datalabels/demo.js:29]: I think this is not an issue. Let's take a look at this demo: https://jsfiddle.net/BlackLabel/jkhx1085/ Until we scroll to the left there are only 3 points (the closest from left + two visible) and we've got two `empty` (`undefined`) values at the start of `series.data` array. I think it is probably related to the necessary calculations between points.

**Comment by TorsteinHonsi** [comment]
Thanks! I have two questions:

1. Doesn't this mean that [Series.prototype.data](https://github.com/highcharts/highcharts/blob/3480cf3f66001bd20d46ad2861ca9d214393cc69/ts/Core/Series/Series.ts#L283) is typed wrong? It should probably be `Array<Point|undefined>`. And if we don't change that type ([and those of the derivatives](https://github.com/highcharts/highcharts/blob/3480cf3f66001bd20d46ad2861ca9d214393cc69/ts/Series/ErrorBar/ErrorBarSeries.ts#L71)) we are destined to make the same mistake again?
2. Instead of making it into `forEach` loops in order to overcome the `undefined` values of `data`, I think it is better that we consistently loop over `series.points`. That is what we usually do, and I suspect that the loops over `series.data` are not really for a good reason. If we change it to loop over `points` and all tests pass, I would say we are safe.

**Comment by bre1470** [comment]
1. You are right. Actually TypeScripts generic type is wrong as any array could contain `undefined`. That it is not included is by design, because the overhead of conditional array access would be significant.
2. Good points!

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**